### PR TITLE
Improve process code in probe

### PR DESCRIPTION
**Pull request by tomwilkie**
- move pidtree into its own module
- desegregate pidtree into tree, walker and reporter
- extend tests
- put command line in the process metadata

might help towards #238

### CODE REVIEW

File: probe/docker/tagger.go [hunk 1]
```diff
@@ -43,7 +43,7 @@ func (t *Tagger) Tag(r report.Report) (report.Report, error) {
 	return r, nil
 }
 
-func (t *Tagger) tag(pidTree tag.PIDTree, topology *report.Topology) {
+func (t *Tagger) tag(pidTree process.Tree, topology *report.Topology) {
```
Comment (paulbellamy) [inline probe/docker/tagger.go]: Consistent terminology, please. Is it a ProcessTree? or a PIDTree?

File: probe/process/reporter.go [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	result.Process.Merge(processes)
+	return result, nil
+}
+
+func (r *Reporter) processTopology() (report.Topology, error) {
+	t := report.NewTopology()
+	err := Walk(r.procRoot, func(p *Process) {
+		pidstr := strconv.Itoa(p.PID)
+		nodeID := report.MakeProcessNodeID(r.scope, pidstr)
+		t.NodeMetadatas[nodeID] = report.NodeMetadata{
```
Comment (paulbellamy) [inline probe/process/reporter.go]: Oh! it's a map. Looks like a struct. Adds a bit of type-safety, but a bit non-obvious.

File: probe/process/walker.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// Hooks exposed for mocking
+var (
+	ReadDir  = ioutil.ReadDir
+	ReadFile = ioutil.ReadFile
+)
+
+// Walk
+var Walk = func(procRoot string, f func(*Process)) error {
```
Comment (paulbellamy) [inline probe/process/walker.go]: Unfortunate that this has to be public.

File: render/detailed_node.go [hunk 4]
```diff
@@ -139,15 +140,17 @@ func addressOriginTable(nmd report.NodeMetadata) (Table, bool) {
...
 	rows := []Row{}
-	if val, ok := nmd["comm"]; ok {
-		rows = append(rows, Row{"Name (comm)", val, ""})
-	}
-	if val, ok := nmd["pid"]; ok {
-		rows = append(rows, Row{"PID", val, ""})
-	}
-	if val, ok := nmd["ppid"]; ok {
-		rows = append(rows, Row{"Parent PID", val, ""})
+	for _, tuple := range []struct{ key, human string }{
```
Comment (paulbellamy) [inline render/detailed_node.go]: nice refactor

File: probe/process/walker.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		pid, err := strconv.Atoi(filename)
+		if err != nil {
+			continue
+		}
+
+		stat, err := ReadFile(path.Join(procRoot, filename, "stat"))
+		if err != nil {
+			continue
+		}
+		splits := strings.Split(string(stat), " ")
```
Comment (paulbellamy) [inline probe/process/walker.go]: strings.Fields?

File: probe/process/walker.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		}
+
+		threads, err := strconv.Atoi(splits[19])
+		if err != nil {
+			return err
+		}
+
+		cmdline := ""
+		if cmdlineBuf, err := ReadFile(path.Join(procRoot, filename, "cmdline")); err == nil {
+			cmdlineBuf = bytes.Replace(cmdlineBuf, []byte{'\000'}, []byte{' '}, -1)
```
Comment (paulbellamy) [inline probe/process/walker.go]: please add test for this

Comment (tomwilkie) [inline probe/process/walker.go]: done.

File: probe/process/reporter.go (hunk 2 above)
Comment (tomwilkie) [inline probe/process/reporter.go]: :-)

File: render/detailed_node.go (hunk 4 above)
Comment (tomwilkie) [inline render/detailed_node.go]: why thank you.

**Comment by paulbellamy** [comment]
LGTM

**Pull request merged**